Marketplace: modernize skeleton loaders

Dan-Q opened the pull request:
### Changes proposed in this Pull Request:

Closes [WCCOM#18034](https://github.com/Automattic/woocommerce.com/issues/18034). Closes [WCCOM#18257](https://github.com/Automattic/woocommerce.com/issues/18257).

This PR replaces all "skeleton" loading indicators on the Marketplace with more-standard ones. It also aims to make those skeleton loaders:
- More-accurately represent this size and shape of the content that will replace them, and
- Refactors the code so that components are responsible for hosting their own skeleton code, attached to an `isLoading` variable, making it easier to stay consistent as changes are made in future

### How to test the changes in this Pull Request:

1. Check out this branch and ensure everything's recompiled with e.g. `cd plugins/woocommerce-admin && pnpm run start:hot`
2. Visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions and observe the new skeleton loaders on the Discover page
3.Visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions&tab=extensions and see the new skeleton loaders on the Browse page (you might also like to do a search)

### Demo video

This demo video artificially slows down loading a _lot_ so the skeletons are very visible:

https://github.com/woocommerce/woocommerce/assets/53293/0b17682b-a87e-4b81-bcaf-abe835be9c5f

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

Improved loading indicators on WooCommerce > Extensions so they more-accurately reflect the layout of the content that will replace them.

</details>

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/category-selector/category-selector.tsx:
```diff
@@ -104,10 +104,13 @@ export default function CategorySelector(): JSX.Element {
-				<Spinner />
-			</div>
+			<ul className="woocommerce-marketplace__category-selector">
+				{[...Array(5)].map(() =>
+					<li className="woocommerce-marketplace__category-item">
+						<CategoryLink slug="" label="" selected={ false } />
+					</li>
+				)}
```
Nice! The attention to detail on even showing a skeleton category selector is ❤️ 

There are some linting issues. Suggestion below.

```suggestion
				{ [ ...Array( 5 ) ].map( () => (
					<li className="woocommerce-marketplace__category-item">
						<CategoryLink slug="" label="" selected={ false } />
					</li>
				) ) }
```

Also, I get a missing key JSX error for the list items. Not sure it's strictly needed in this particular case.

Dan-Q replied:
Yeah, I flip-flopped on adding a key. Obviously it's not needed because we don't mind if the entire container needs re-rendering on a change (the only change that can happen requires a re-render of the container _anyway_!), but if the warning bugs people we could fix it...

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -5,6 +5,50 @@
+			.woocommerce-marketplace__product-card__description,
+			.woocommerce-marketplace__product-card__price {
+				animation: loading-fade 1.6s ease-in-out infinite;
+				background: repeating-linear-gradient(#f0f0f0, #f0f0f0 13px, transparent 13px, transparent 21px);
```
This is magical. Love some clever gradient-based CSS techniques.

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/product-loader/product-loader.tsx:
```diff
@@ -5,15 +5,25 @@
+	placeholderCount?: number;
+}
+
+export default function ProductLoader(props: ProductLoaderProps): JSX.Element {
```
Linting issue

```suggestion
export default function ProductLoader(
	props: ProductLoaderProps
): JSX.Element {
```

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/product-loader/product-loader.tsx:
```diff
@@ -5,15 +5,25 @@
-				<div className="woocommerce-marketplace__product-loader-divider divider-2"></div>
-				<div className="woocommerce-marketplace__product-loader-divider divider-3"></div>
+		<div className="woocommerce-marketplace__product-list">
+			{ hasTitle !== false &&
+				<ProductListHeader title="" groupURL={ null } />
+			}
+			<div className="woocommerce-marketplace__product-list-content">
+				{[...Array(placeholderCount)].map(() =>	<ProductCard />	)}
```
Linting errors suggested below. Also seeing the missing key JSX error here too.

```suggestion
			{ hasTitle !== false && (
				<ProductListHeader title="" groupURL={ null } />
			) }
			<div className="woocommerce-marketplace__product-list-content">
				{ [ ...Array( placeholderCount ) ].map( () => (
					<ProductCard />
				) ) }
```

kdevnel reviewed (dismissed):
Thanks for this Dan! Looks amazing. The detail of these skeletons is phenomenal. Love your approach. 

Added some linting suggestions (they may not be exhaustive so I'd run the local linter to be sure) and some other JSX errors.

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/category-selector/category-selector.tsx:
```diff
@@ -104,10 +104,13 @@ export default function CategorySelector(): JSX.Element {
+			<ul className="woocommerce-marketplace__category-selector">
+				{[...Array(5)].map(() =>
+					<li className="woocommerce-marketplace__category-item">
+						<CategoryLink slug="" label="" selected={ false } />
```
What if we had a dedicated loading prop?
This works ok, but it's not clear how to start the loader.
Definitely not a blocker.

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -5,6 +5,50 @@
+			}
+
+			.woocommerce-marketplace__product-card__icon {
+				background: #f0f0f0;
```
`#f0f0f0` could be a variable.

Dan-Q replied:
It absolutely should!

I had a mental note that I should do this. Then I didn't. It was getting late in the day and I was running around trying to make dog/child things happen! (Also why I didn't run the linter, oops!)

bgrgicak reviewed (commented):
@Dan-Q where can I see the design for skeleton loaders?
They seem to work, but I'm not sure if they look as expected.

kdevnel commented:
> @Dan-Q where can I see the design for skeleton loaders? They seem to work, but I'm not sure if they look as expected.

@bgrgicak that's a great point. I don't think we have designs for this, since we're defining loading states as we go. It's probably a good call to get @poligilad-auto involved to have a look and share any thoughts.

bgrgicak commented:
>  I don't think we have designs for this, since we're defining loading states as we go.

That's ok, a screenshot in the PR description would also work.

poligilad-auto commented:
Hey @Dan-Q,
Design wise, we should match the [existing components](https://woocommerce.github.io/woocommerce-admin/#/components/) like TablePlaceholder, which I think you are already doing?

Also, I'm wondering if we need to reproduce all of the existing elements, normally placeholders render less components than the actual component, not to overwhelm the screen when there is no need.

Here is a Youtube example 👇
1. The page only renders the _video placeholder_ and not the _shorts placeholder_.
2. Notice that the actual video placeholder component just has two main "lines" it doesn't render an extra line for the smaller info below the main title.
<img width="2124" alt="image" src="https://github.com/woocommerce/woocommerce/assets/83961704/e7646bc7-7ee4-466e-a9f3-b7ec54ed6bc1">

Maybe in our case we could determine which elements to render for the _placeholder component_? Is this too much work?
Could you share a screenshot so it's easier for me to view?

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -5,6 +5,58 @@
+			.woocommerce-marketplace__product-card__vendor,
+			.woocommerce-marketplace__product-card__description,
+			.woocommerce-marketplace__product-card__price {
+				height: calc(13px * 1.5);
```
I like to store values like these in variables so that there is at least some description to what they mean.
Otherwise, we can just have the final value.

andfinally replied:
I tried this, but `calc` didn't like calculating with a var in the other place we use this value, so I just added a couple of comments.

andfinally reviewed (approved):
Looks lovely, thanks! Pushed a couple of small tweaks in `3ab2db07cec4cdb46d1cec112318044ebaa1e875` and `f5725288cd175dfdf1e8880e5a3d6358ba57acc9` – when you call `Array.map` the index is the second param. 

If the tests pass I might have a go at squashing this PR and merging.

andfinally committed:
This branch replaces all "skeleton" loading indicators on the Marketplace with more-standard ones. It also:

- Aims to make those skeleton loaders more accurately represent this size and shape of the content that will replace them.
- Refactors the code so that components are responsible for hosting their own skeleton code, attached to an `isLoading` variable, making it easier to stay consistent as changes are made in future.
